### ISSUE TITLE
upload events not working

### ISSUE THREAD

**Issue by madovsky** [i-rinat/freshplayerplugin#308]
I noticed that the events from fileListener upload class does not fire,
unless the OnCompleteData.
for example a developed a progress bar which works on flash player NPAPI
and Chrome PPAPI but not with freshwrapper.
Rinat, you can test it on the file manager on your Teacher control panel (account I sent to you)

**Comment by i-rinat** [comment]
https://github.com/i-rinat/freshplayerplugin/blob/cecfd52016fd0e31c7346bdaa1eb6ecb85c01d0d/src/ppb_url_loader.c#L472

It was so long ago, I forgot :)
NPAPI treats file upload as an atomic request. I wonder, how NPAPI Flash manages to get upload progress.

**Comment by madovsky** [comment]
TODO list never ends ;)...
you mean "how PPAPI..." and not "how NPAPI"....?

**Comment by i-rinat** [comment]
> you mean "how PPAPI..." and not "how NPAPI"....?

Nope. I meant NPAPI Flash.
NPAPI allows plugin to make requests. Browser then returns so called stream. It calls plugin after another set of data arrives. So it's possible to get download progress. But when you want to upload data, plugin must create a file with request content (if it's a POST request), and then a browser makes a request. But there is no way to figure out how much of data was uploaded. All plugin gets is a notification, when servers responds. With large uploads that could take a while.

I've investigated a bit into Flash 11.2. It doesn't use browser facilities, but makes own connection.

**Comment by madovsky** [comment]
ha ok, yes FileReference uses its own connection for sure...

**Comment by i-rinat** [comment]
PPB_URLLoader is the most trickiest part, indeed. It has a lot of code around `NPP_*` functions as well. There also were many subtle bugs which are now eliminated, I hope. That's why I'm afraid of adding another code path for making connections not through the browser. It's fragile. Also working through proxies will be broken in a weird way. Some request will work, while others not.

**Comment by madovsky** [comment]
is PPB_ pepperflash and NPP_ NPAPI flash?
if FileReference has it's own connection does it mean you cannot have access to it?

**Comment by i-rinat** [comment]
> is PPB_ pepperflash and NPP_ NPAPI flash?

`PP` means Pepper Plugin. `B` means Browser. All interfaces provided by a browser have name starting with `PPB_`. API also defines interfaces which are provided by plugin. Their names start with `PPP_`, `P` for Plugin.

NPAPI have similar naming. (Or more likely PPAPI have similar to NPAPI naming scheme). Browser function start with `NPN_`, `NP` for Netscape Plugin, and `N` for Navigator. There are also functions, which plugin must implement. Their names start with `NPP_`. With, similarly, `P` for Plugin.

> if FileReference has it's own connection does it mean you cannot have access to it?

No, I don't have any access to AS3 functions at all. Flash **is** the abstraction layer. AS3 programs doesn't know about platform details they are working on, Flash translates class method calls into platform-specific calls, which are PPAPI in this case.

**Comment by madovsky** [comment]
ok I see, thanks for the explanation. If I understand well so it must create a NPN function that bridge the PPB one?

**Comment by i-rinat** [comment]
> If I understand well so it must create a NPN function that bridge the PPB one?

There are already predefined set of `NPN_*` functions, one cannot add own. The list of functions was [here](https://developer.mozilla.org/en-US/docs/Plugins/Guide/Browser_Side_Plug-in_API), but it looks like documentation is broken at the moment. (Prototypes of those functions are still in [npapi.h](https://dxr.mozilla.org/mozilla-central/source/dom/plugins/base/npapi.h), but that's still not a proper documentation).

In general, freshplayerplugin acts like a normal NPAPI plugin. That means, it can use browser part of NPAPI, but also can do platform-related actions directly. And if I understand your question right, you are asking how to implement upload progress notification. Answer is — to make requests not through browser part of NPAPI, but directly, calling `socket()`, `connect()`, and then `send()`/`recv()`. That approach is already used for `PPB_TCPSocket` and `PPB_UDPSocket`, but there're no choice, since NPAPI doesn't have corresponding functions.

Contrary, there is API for making HTTP/HTTPS requests and freshwrapper currently uses it. Here comes a tricky part. Combining two approaches will increase code complexity and definitely introduce subtle errors. Also proxying will be broken for POST requests. Another solution is to ditch request-related NPAPI part completely, and reimplement request from scratch. Resulting code will be simpler and less error-prone. But then proxying will be broken for all requests.

Also HTTPS issue arises. It will be required to decide whenever freshwrapper checks HTTPS certificates validity or not. Now it's browser concern, but user can override security checks. How to manage such exceptions in freshwrapper? Allow everything or develop a way to add exceptions? I don't know.

**Comment by madovsky** [comment]
thanks for your explanation, I found these links
https://groups.google.com/forum/#!topic/firebreath-dev/Sf5IBduRE2Y
https://developer.chrome.com/apps/npapi
https://support.google.com/chrome/answer/6213033?hl=en
https://www.chromium.org/developers/npapi-deprecation
http://www.firebreath.org/display/documentation/FireBreath+2.0%3A+Browser+Plugins+in+a+post-NPAPI+world
https://github.com/kilfu0701/npapi-thk-nico-dl-linux

**Comment by madovsky** [comment]
found this link too
http://peter.sh/experiments/chromium-command-line-switches/

**Comment by madovsky** [comment]
Rinat, I saw that you can access to javascript from NPAPI,
so maybe it's possible to use some JS lib to check the progress since the start and complete
are successfully detected

**Comment by i-rinat** [comment]
Yeah, looks like there are progress events in XMLHttpRequest for uploads too.

**Comment by i-rinat** [comment]
XMLHttpRequest can't be used to fully implement uploads (or downloads). There are limitation for Javascript code which prevents it from making requests to other domains. Flash implements own security mechanism, by implicitly requesting `crossdomain.xml` file. So, that's a dead end.
